File: packages/shared/src/types/userContext.ts
/**
 * The user's "Believe Profile" — who they are, what they offer, and who
 * they're trying to reach. Optional context fed into every AI generation
 * and personalization call so output can reference the sender's real
 * situation instead of staying generic.
 */
export interface UserContext {
  userId: string;
  aboutMe: string | null;
  companyInfo: string | null;
  servicesOrProducts: string | null;
  skillsAndExperience: string | null;
  achievements: string | null;
  targetAudience: string | null;
  updatedAt: string;
}

/**
 * Flattens a UserContext into a single string suitable for dropping into an
 * AI prompt as background context. Skips empty fields rather than emitting
 * "About me: " noise.
 */
export function formatUserContextForPrompt(context: UserContext | null): string | undefined {
  if (!context) return undefined;
  const parts = [
    context.aboutMe && `About the sender: ${context.aboutMe}`,
    context.companyInfo && `Company: ${context.companyInfo}`,
    context.servicesOrProducts && `Services/products: ${context.servicesOrProducts}`,
    context.skillsAndExperience && `Skills/experience: ${context.skillsAndExperience}`,
    context.achievements && `Achievements: ${context.achievements}`,
    context.targetAudience && `Target audience: ${context.targetAudience}`,
  ].filter(Boolean);
  return parts.length > 0 ? parts.join("\n") : undefined;
}

File: apps/backend/utils/user_context.py
"""Flattens a stored Believe Profile document into a prompt-ready string —
mirrors packages/shared/src/types/userContext.ts's formatUserContextForPrompt
so both services describe a sender identically.
"""

from __future__ import annotations


def format_user_context_for_prompt(context: dict | None) -> str | None:
    if not context:
        return None
    candidates: list[str | None] = [
        context.get("aboutMe") and f"About the sender: {context['aboutMe']}",
        context.get("companyInfo") and f"Company: {context['companyInfo']}",
        context.get("servicesOrProducts") and f"Services/products: {context['servicesOrProducts']}",
        context.get("skillsAndExperience") and f"Skills/experience: {context['skillsAndExperience']}",
        context.get("achievements") and f"Achievements: {context['achievements']}",
        context.get("targetAudience") and f"Target audience: {context['targetAudience']}",
    ]
    parts: list[str] = [p for p in candidates if p]
    return "\n".join(parts) if parts else None
